File: interface/buttoncell.cpp
#include "buttoncell.h"

ButtonCell::ButtonCell(QToolButton *parent) :
    QToolButton(parent)
{
    connect(this, &ButtonCell::clicked, this, &ButtonCell::onLetterClicked);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
    currentMenu_ = createMenu();
    setMenu(currentMenu_);
}

void ButtonCell::connectToPanel(QObject *gb)
{
    connect(this, SIGNAL(onLetterChosen(QChar)), gb, SLOT(onCellChosen(QChar)));
    connect(this, SIGNAL(onLetterPushed()), gb, SLOT(onCellPushed()));
}

QMenu* ButtonCell::createMenu() {
    signalMapper_ = new QSignalMapper(this);
    QMenu* menu = new QMenu(tr("Set of letters"));
    for(wchar_t c = L'а'; c <= L'я'; ++c){
        actions_.push_back(new QAction(QString(c), this));
        menu->addAction(actions_.back());
        signalMapper_->setMapping(actions_.back(), QString(c));
        connect(actions_.back(), SIGNAL(triggered()), signalMapper_, SLOT(map()));
    }
    connect(signalMapper_, SIGNAL(mapped(QString)), this, SLOT(setLetter(QString)));
    return menu;
}

File: python-src/interface/ButtonCell.py
__author__ = 'akhtyamovpavel'

from PySide import QtCore, QtGui

class ButtonCell(QtGui.QToolButton):
    on_letter_chosen = QtCore.Signal(str)
    on_letter_pushed = QtCore.Signal()


    @QtCore.Slot(str)
    def set_letter(self, letter):
        self.on_letter_chosen.emit(letter)

    @QtCore.Slot()
    def on_letter_clicked(self):
        self.on_letter_pushed.emit()

    def __init__(self, language):
        super(ButtonCell, self).__init__()
        self.__language__ = language
        self.clicked().connect(self.on_letter_clicked)

        self.setSizePolicy(QtGui.QSizePolicy.Expanding, QtGui.QSizePolicy.Preferred)
        self.create_menu()


    def create_menu(self):
        self.__signal_mapper__ = QtCore.QSignalMapper(self)
        self.__menu__ = QtGui.QMenu('Set of letters')

        self.__actions__ = list()

        current_language = self.__language__.get_list()

        for letter in current_language:
            self.__actions__.append(QtGui.QAction(letter, self))
            self.__menu__.addAction(self.__actions__[-1])
            self.__signal_mapper__.setMapping(self.__actions__[-1], letter)
            self.__actions__[-1].triggered.connect(self.__signal_mapper__.map)

        self.__signal_mapper__.mapped().connect(self.set_letter)


    def connect_to_panel(self, gb):
        pass

